From 6dc75545fa10b4bc2c3653c02f68b76394bbbb5d Mon Sep 17 00:00:00 2001
From: Steve deRosier <steve.derosier@lairdtech.com>
Date: Wed, 17 Feb 2016 14:58:50 -0800
Subject: [PATCH] laird: ath6kl: ignore WMI_TXE_NOTIFY_EVENTID based on fw
 capability flags

The 6004 redefines the WMI_TXE_NOTIFY_EVENTID event number and sends the
new event frequently. However it doesn't have the tx-err-notify feature
and thus this firmware capability flag isn't set on the firmware package.
By guarding the processing of this event by the same method we guard the
sending of the WMI_SET_TXE_NOTIFY_CMDID command, we can ignore the spurious
event that we don't know how to process yet.

Without this change we call cfg80211_cqm_txe_notify() with possibly bad
data.

Bug: 9049
---
 drivers/net/wireless/ath/ath6kl/wmi.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/drivers/net/wireless/ath/ath6kl/wmi.c b/drivers/net/wireless/ath/ath6kl/wmi.c
index 096dd7f5ccbc..f471309d4db4 100644
--- a/drivers/net/wireless/ath/ath6kl/wmi.c
+++ b/drivers/net/wireless/ath/ath6kl/wmi.c
@@ -1661,6 +1661,10 @@ static int ath6kl_wmi_txe_notify_event_rx(struct wmi *wmi, u8 *datap, int len,
 	if (len < sizeof(*ev))
 		return -EINVAL;
 
+	if ( vif->nw_type != INFRA_NETWORK ||
+	    !test_bit(ATH6KL_FW_CAPABILITY_TX_ERR_NOTIFY, vif->ar->fw_capabilities))
+		return -EOPNOTSUPP;
+
 	if (vif->sme_state != SME_CONNECTED)
 		return -ENOTCONN;
 
-- 
2.11.0

